### ISSUE TITLE
mini-RFC: add 'compile_size' method to FontWrite trait

### ISSUE THREAD

**Issue by cmyr** [googlefonts/fontations#543]
Several times now I've run into situations where it would be helpful to know how many bytes the compiled representation of a given table will require, before actually compiling it.

This could potentially be helpful for things like table splitting; currently we do splitting by manual byte-mucking after attempting an initial compilation. With this change it would become possible for us to determine which splits are required ahead of time, which would let us implement these algorithms entirely in write-fonts, which would be a cleaner and more idiomatic solution.

One problem with this idea is that the sizes that we would calculate would not account for the table-packing algorithm's ability to deduplicate identical tables, which can significantly decrease table sizes in some cases. In effect, the size returned by this method would represent an upper bound. I think this is still useful, since in cases such as table splitting the only consequence might be that we unnecessarily split tables that are very near the overflow limit anyway, which seems okay; we would still be producing valid output.

This should also be quite easy to implement, since we already need to know how to calculate the sizes and lengths of everything in codegen, in order to correctly parse tables.

The main downsides is that this is not 100% necessary work, and it adds code that we have to compile and maintain.

**Comment by rsheeter** [comment]
A concrete example would help.

**Comment by cmyr** [comment]
I just remembered one of the concrete motivations for this idea: in several places we need to choose between two different formats for a given subtable (specifically for contextual lookups) and currently the only real way to do this is to actually generate and then serialize the different options, in order to determine their final sizes. In theory it would be nice if we could skip that last step, since it involves running the full packing algorithm.

**Comment by anthrotype** [comment]
in theory it would be nice to have, but in practice i'm curious how can you know how many bytes something will take before you actually serialize (run the full packing on) it?

**Comment by cmyr** [comment]
correct, this method would produce an 'upper bound' on size, since it could not perform deduplication. However for many of the cases that we care about I think duplicate objects would be unlikely or uncommon, and this would still be useful.
